### PR TITLE
[Merged by Bors] - Add PBR textures

### PR DESCRIPTION
**Pull request by mtsr**
This PR adds normal maps on top of PBR #1554. Once that PR lands, the changes should look simpler.

Edit: Turned out to be so little extra work, I added metallic/roughness texture too. And occlusion and emissive.

### CODE REVIEW

File: crates/bevy_render/src/shader/shader_reflect.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ impl ShaderLayout {
                 for input_variable in module.enumerate_input_variables(None).unwrap() {
                     if input_variable.name == GL_VERTEX_INDEX
                         || input_variable.name == GL_INSTANCE_INDEX
+                        || input_variable.name == GL_FRONT_FACING
                     {
                         continue;
                     }
```
Comment (mtsr) [inline crates/bevy_render/src/shader/shader_reflect.rs:39]: I might misunderstand the point of this condition, if so I'd love to learn. But wouldn't a more long-term solution be (snipped some `use` changes for clarity):

```diff
diff --git a/crates/bevy_render/src/shader/shader_reflect.rs b/crates/bevy_render/src/shader/shader_reflect.rs
index 238da1ef..95761fed 100644
--- a/crates/bevy_render/src/shader/shader_reflect.rs
+++ b/crates/bevy_render/src/shader/shader_reflect.rs
@@ -31,8 +32,9 @@ impl ShaderLayout {
                 // obtain attribute descriptors from reflection
                 let mut vertex_attributes = Vec::new();
                 for input_variable in module.enumerate_input_variables(None).unwrap() {
-                    if input_variable.name == GL_VERTEX_INDEX
-                        || input_variable.name == GL_INSTANCE_INDEX
+                    if input_variable
+                        .decoration_flags
+                        .contains(ReflectDecorationFlags::BUILT_IN)
                     {
                         continue;
                     }
```

**Comment by mtsr** [comment]
https://user-images.githubusercontent.com/1319584/110938730-9964f300-8334-11eb-80d8-320bf661ef3a.mp4

File: crates/bevy_render/src/shader/shader_reflect.rs (hunk 1 above)
Comment (cart) [reply crates/bevy_render/src/shader/shader_reflect.rs:39]: Oooh that does seem like a win / appears to do what we want it to.

**Comment by mtsr** [comment]
Rebased on pbr.

**Comment by mtsr** [comment]
Rebased on the latest pbr.

**Comment by mtsr** [comment]
And rebased onto main now that #1554 has been merged.

File: crates/bevy_render/src/mesh/mesh.rs [hunk 2]
```diff
@@ -237,6 +237,9 @@ impl Mesh {
     /// The direction the vertex normal is facing in.
     /// Use in conjunction with [`Mesh::set_attribute`]
     pub const ATTRIBUTE_NORMAL: &'static str = "Vertex_Normal";
+    ///
```
Comment (jakobhellermann) [inline crates/bevy_render/src/mesh/mesh.rs:240]: I think you forgot to add the doc comment here

File: crates/bevy_pbr/src/render_graph/pbr_pipeline/pbr.frag [hunk 3]
```diff
@@ -258,10 +290,48 @@ void main() {
...
+    float occlusion = metallic_roughness.r;
+#    elif defined(STANDARDMATERIAL_OCCLUSION_TEXTURE)
+    float occlusion = texture(sampler2D(StandardMaterial_occlusion_texture, StandardMaterial_occlusion_texture_sampler), v_Uv).r;
+#    else
+    float occlusion = 1.0;
+#    endif
+
+#    ifdef STANDARDMATERIAL_EMISSIVE_TEXTURE
+    // TODO use .a for exposure compensation in HDR
+    vec3 emissive.rgb *= texture(sampler2D(StandardMaterial_emissive_texture, StandardMaterial_emissive_texture_sampler), v_Uv).rgb;
```
Comment (jakobhellermann) [inline crates/bevy_pbr/src/render_graph/pbr_pipeline/pbr.frag:332]: the `vec3` should go, otherwise the shader fails to compile.

**Comment by jakobhellermann** [comment]
Should the standard shapes like cube or sphere provide the `tangent` vertex attribute?

File: crates/bevy_pbr/src/material.rs [hunk 4]
```diff
@@ -24,7 +24,25 @@ pub struct StandardMaterial {
...
+    pub normal_map: Option<Handle<Texture>>,
+    #[render_resources(ignore)]
+    #[shader_def]
+    pub double_sided: bool,
+    #[render_resources(ignore)]
+    #[shader_def]
+    pub occlusion_shares_metallic_roughness_texture: bool,
+    #[shader_def]
+    pub occlusion_texture: Option<Handle<Texture>>,
+    // Use a color for user friendliness even though we technically don't use .a
```
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/material.rs:40]:
```suggestion
    // Use a color for user friendliness even though we technically don't use the alpha channel
```

File: crates/bevy_pbr/src/render_graph/pbr_pipeline/pbr.frag [hunk 5]
```diff
@@ -310,7 +380,9 @@ void main() {
     vec3 diffuse_ambient = EnvBRDFApprox(diffuseColor, 1.0, NdotV);
     vec3 specular_ambient = EnvBRDFApprox(F0, perceptual_roughness, NdotV);
 
-    output_color.rgb = light_accum + (diffuse_ambient + specular_ambient) * AmbientColor;
+    output_color.rgb = light_accum;
```
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/render_graph/pbr_pipeline/pbr.frag:383]: Tiny comments here would make it easier for folks new to rendering to figure out exactly what's happening here.

See the `// tone_mapping` comment just below.

File: examples/3d/pbr_textures.rs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        })
+        .spawn(PerspectiveCameraBundle {
+            transform: Transform::from_xyz(0.7, 0.7, 1.0)
+                .looking_at(Vec3::new(0.0, 0.3, 0.0), Vec3::Y),
+            ..Default::default()
+        });
+}
+
+/// this component indicates what entities should rotate
+struct Rotator;
```
Comment (alice-i-cecile) [inline examples/3d/pbr_textures.rs:58]:
```suggestion
struct Rotates;
```

IMO adjectives are generally much clearer than nouns for the name of components like this.

File: examples/README.md [hunk 7]
```diff
@@ -88,6 +88,7 @@ Example | File | Description
 `orthographic` | [`3d/orthographic.rs`](./3d/orthographic.rs) | Shows how to create a 3D orthographic view (for isometric-look games or CAD applications)
 `parenting` | [`3d/parenting.rs`](./3d/parenting.rs) | Demonstrates parent->child relationships and relative transformations
 `pbr` | [`3d/pbr.rs`](./3d/[pbr].rs) | Demonstrates PBR properties Roughness/Metallic
```
Comment (alice-i-cecile) [inline examples/README.md:90]:
```suggestion
`pbr` | [`3d/pbr.rs`](./3d/[pbr].rs) | Demonstrates use of Physically Based Rendering (PBR) properties
```

Cleanup for novice-clarity and reduced description churn.

File: examples/README.md [hunk 8]
```diff
@@ -88,6 +88,7 @@ Example | File | Description
...
+`pbr_textures` | [`3d/pbr_textures.rs`](./3d/[pbr].rs) | Demonstrates PBR textures (normals/roughness/metallic/occlusion)
```
Comment (alice-i-cecile) [inline examples/README.md:91]:
```suggestion
`pbr_textures` | [`3d/pbr_textures.rs`](./3d/[pbr].rs) | Demonstrates setting PBR properties as a texture, allowing you to vary the properties across your model
```

Clearer explanation for beginners.

**Comment by pkupper** [comment]
Looking at the image below, there is clearly something wrong with the normal mapping in the example.
bevy_normalmapping
One possible mistake I found is that in the fragment shader the sample taken from the normal map is not converted into the -1.0 to 1.0 range before being transformed to world space. But that alone doesn't seem to fix the issue.

**Comment by alice-i-cecile** [comment]
Transferring @pkupper's comment from [Discord](https://discord.com/channels/691052431525675048/743663924229963868/824398688105267271) for discoverability:

> Another thing that's missing is calculating normals and tangents if they are not present in the gltf. Tried a normal map sample gltf for easier debugging (https://github.com/KhronosGroup/glTF-Sample-Models) and it didn't load due to missing tangents. The gltf spec defines how both should be calculated (https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#meshes). But that is probably something for another PR.

**Comment by mtsr** [comment]
> Should the standard shapes like cube or sphere provide the `tangent` vertex attribute?

Possibly. Do we want to pull that into this PR? (neutral question) It has both pros and cons, I think, but I guess it does make sense to make this more fit for general use, rather than having to add that in a separate PR.

**Comment by mtsr** [comment]
We're hitting a minor snag, where `GltfLoader` passes off the textures to be loaded as a `Vec<AssetPath>`, without any texture information. Specifically, we cannot tell `AssetServer` to create Linear textures from these, rather than Srgb.

I don't really know where to start fixing this. If somebody is able to help out, so I don't have to also dig into the Asset system, please do. Otherwise I may look into it this weekend or next week.

**Comment by jakobhellermann** [comment]
I think a second PR is fine for adding the missing tangent vertex attributes.

**Comment by mtsr** [comment]
Rebase on #1762 in progress...

**Pull request closed without merging**